### DISCUSSION TITLE
[bug] `default_factory` called twice when parse input by FastAPI/route

### DISCUSSION THREAD

**Discussion by dacrystal** [fastapi/sqlmodel#608] [Show and tell]
## Current Behavior (example Code):
```py
from typing import Optional
from sqlmodel import Field, SQLModel


def gen_id():
    gen_id.i += 1
    print(f"gen_id called. id={gen_id.i}")
    return gen_id.i


gen_id.i = 0


class Hero(SQLModel):
    id: Optional[int] = Field(primary_key=True, default_factory=gen_id)
    name: str
    secret_name: str
    age: Optional[int] = None


hero_1 = Hero.validate({"name": "Deadpond",
                        "secret_name": "Dive Wilson"})


assert hero_1.id == 1
```

## Expected Behavior

`default_factor` and `validations` should run once.

## Root cause
The issue is that `validation_model` called twice directly in `L589`  and indirectly in `L592` (__init__)
https://github.com/tiangolo/sqlmodel/blob/43a689d369f52b72aac60efd71111aba7d84714d/sqlmodel/main.py#L583-L605

in `__init__` :

https://github.com/tiangolo/sqlmodel/blob/43a689d369f52b72aac60efd71111aba7d84714d/sqlmodel/main.py#L498-L498

**Comment by antont** [comment]
Just a note, that I guess that with the _multiple models pattern_ this does not typically happen, because the *Create model does not include fields with factories, but typically the db model does. https://sqlmodel.tiangolo.com/tutorial/fastapi/multiple-models/

Sure would be nice to fix anyhow.

**Reply by dacrystal** [reply]
Yes multiple-models-pattern (what I'm using) does fix it for `default_factory`.
However, validation still run twice for the same data when used with FastAPI(fast/high-performance) or when validating a model (for some hypothetical use case) in general.

**Comment by dacrystal** [comment]
@antont 
I was wrong for multiple-models-pattern, it seems it didn't fix it (I was testing the wrong endpoint earlier).

`from_orm` and `parse_as_obj` still cause validations called twice.

Here is a more comprehensive  example:

```py
from typing import List, Optional
from sqlmodel import Field, SQLModel
from pydantic import parse_obj_as


def gen_id():
    gen_id.i += 1
    print(f"gen_id called. id={gen_id.i}")
    return gen_id.i


gen_id.i = 0


class HeroBase(SQLModel):
    name: str = Field(index=True)
    secret_name: str
    age: Optional[int] = Field(default=None, index=True)


class Hero(HeroBase, table=True):
    id: int = Field(primary_key=True, default_factory=gen_id)


class HeroCreate(HeroBase):
    pass


hero_create = HeroCreate(name="Deadpond", secret_name="Dive Wilson")
hero_create_list = [
    HeroCreate(name="Spider-Boy", secret_name="Pedro Parqueador"),
    HeroCreate(name="Deadpond", secret_name="Dive Wilson")]


# This work, but...
hero_parse_obj = Hero.parse_obj(hero_create)

gen_id.i = 0
# default_factory/validations called twice
hero_validate = Hero.validate({"name": "Deadpond",
                               "secret_name": "Dive Wilson"})

# default_factory/validations called twice
gen_id.i = 0
hero_from_orm = Hero.from_orm(hero_create)

# default_factory/validations called twice
gen_id.i = 0
hero_list = parse_obj_as(List[Hero], hero_create_list)


assert hero_parse_obj.id == 1
assert hero_validate.id == 1
assert hero_from_orm.id == 1
assert hero_list[0].id == 1

```


Or I might be doing it wrongly.

**Reply by antont** [reply]
You are probably right, I guess in our case we just haven't noticed that, with DB generated IDs.
